python/foglamp/services/core/api/filters.py

```diff
@@ -0,0 +1,270 @@
+                            "', doesn't match the specified one '" + \
+                            plugin_name + "', type 'filter'"
+            raise ValueError(error_message)
+
+        #################################################
+        # Set string value for 'default' if type is JSON
+        # This is required by the configuration manager
+        ################################################# 
+        for key, value in plugin_config.items():
+            if value['type'] == 'JSON':
+                value['default'] = json.dumps(value['default'])
+
+        await cf_mgr.create_category(category_name=filter_name,
+                                     category_description=filter_desc,
+                                     category_value=plugin_config)
+
+        # Fetch the new created filter: get category items
+        category_info = await cf_mgr.get_category_all_items(category_name=filter_name)
+        if category_info is None:
+            message = "No such '%s' filter found" % filter_name
+            raise ValueError(message)
+        else:
+            # Success: return new filter content
+            return web.json_response({'filter': filter_name,
+                                      'description': filter_desc,
+                                      'value': category_info})
+
+    except ValueError as ex:
+        _LOGGER.exception("Add filter, caught exception: " + str(ex))
+        raise web.HTTPNotFound(reason=str(ex))
+    except Exception as ex:
+        _LOGGER.exception("Add filter, caught exception: " + str(ex))
+        raise web.HTTPInternalServerError(reason=str(ex))
+
+"""
+    Add filter names to "filter" item in {service_name}
+
+    PUT /foglamp/filter/{service_name}/pipeline
+ 
+    'pipeline' is the array of filter category names to set
+    into 'filter' default/value properties
+
+    :Example: set 'pipeline' for service 'NorthReadings_to_PI'
+    curl -X PUT http://localhost:8081/foglamp/filter/NorthReadings_to_PI/pipeline -d 
+    '{
+        "pipeline": ["Scale10Filter", "Python_assetCodeFilter"],
+    }'
+
+    Configuration item 'filter' is added to {service_name}
+    or updated with the pipeline list
+
+    Returns the filter pipeline on success:
+    {"pipeline": ["Scale10Filter", "Python_assetCodeFilter"]} 
+
+    Query string parameters:
+    - append_filter=true|false       Default true
+    - allow_duplicates=true|false    Default true
+
+    :Example:
+    curl -X PUT http://localhost:8081/foglamp/filter/NorthReadings_to_PI/pipeline?append_filter=true|false -d
+    '{
+        "pipeline": ["Scale10Filter", "Python_assetCodeFilter"],
+    }'
+    curl -X PUT http://localhost:8081/foglamp/filter/NorthReadings_to_PI/pipeline?allow_duplicates=true|false -d
+    '{
+        "pipeline": ["Scale10Filter", "Python_assetCodeFilter"],
+    }'
+
+    NOTE: the method also adds the filters category names under
+    parent category {service_name}
+"""
```

Comment:
Move docstring inside for add_filters_pipeline def

Reply:
The comment is not clear

Reply:
ok, got it